src/workspaceState.ts

```diff
@@ -0,0 +1,93 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { WorkspaceStatus } from '@gitpod/public-api/lib/gitpod/experimental/v1';
+import { WorkspaceInstanceStatus_Phase } from '@gitpod/public-api/lib/gitpod/experimental/v1/workspaces_pb';
+import * as vscode from 'vscode';
+import { Disposable } from './common/dispose';
+import Log from './common/logger';
+import { GitpodPublicApi } from './publicApi';
+
+export class WorkspaceState extends Disposable {
+
+    static POLL_INTERVAL = 5000;
+
+    private poolHandle: NodeJS.Timer | undefined;
+
+    private workspaceStatePromiseResolver!: () => void;
+    readonly workspaceStatePromise = new Promise<void>((r) => this.workspaceStatePromiseResolver = r);
+    private workspaceState: WorkspaceStatus | undefined;
+
+    private _onWorkspaceStatusChanged = this._register(new vscode.EventEmitter<void>());
+    readonly onWorkspaceStatusChanged = this._onWorkspaceStatusChanged.event;
+
+    constructor(
+        readonly workspaceId: string,
+        private readonly publicApi: GitpodPublicApi,
+        private readonly logger: Log,
+    ) {
+        super();
+
+        this.logger.trace(`WorkspaceState manager for workspace ${workspaceId} started`);
+
+        // this._register(this.publicApi.onWorkspaceStatusUpdate(u => this.checkWorkspaceState(u)));
+        // this.publicApi.startWorkspaceStatusStreaming(workspaceId);
+
+        this.pollWorkspaceState();
+        this.poolHandle = setInterval(() => {
+            this.pollWorkspaceState();
+        }, WorkspaceState.POLL_INTERVAL);
```

Comment:
question: is it correct that here it can happen that before we can display our workspace stopping dialog, the VS Code SSH dialog will show first?

Reply:
yes, if remote-ssh extension detects connection is lost first it will show before our dialog